### ISSUE TITLE
[Perf]: Severe NVFP4 decode throughput regression on Blackwell (B300/GB300): uninitialized swizzled scale buffer in `create_fp4_scale_tensor`

### ISSUE THREAD

**Issue by qiching** [vllm-project/vllm#45741]
## Summary

On Blackwell GPUs (B300 and GB300), NVFP4 (`--quantization modelopt_fp4`) decode throughput regressed by **~87–93%** for multiple models. The model output also degenerates: the generated sequence length collapses far below the requested value, and measured inter-token latency rises ~8–15×.

We bisected the functional cause to a single line in `vllm/_custom_ops.py::create_fp4_scale_tensor`, where the **swizzled** scale-factor buffer allocation was changed from `torch.zeros` to `torch.empty` (introduced by [vllm-project/vllm#42988](https://github.com/vllm-project/vllm/pull/42988) "[Perf] zeros -> empty to remove additional fill", commit `b29cbf0652`). The swizzled scale buffer is padded, and the NVFP4 quantization kernel does not write every padded element that the downstream NVFP4 GEMM reads. With `torch.empty`, those padded scale factors are left uninitialized, which corrupts dequantization.

Reverting that one allocation back to `torch.zeros` fully restores throughput, inter-token latency, and output length: on B300 (Llama-3.3-70B, con=1) and on GB300 (both a dense 70B and a 1T-A32B MoE, con=1); and on GB300 the fixed build additionally scales cleanly across concurrencies 1/16/64/128.

## Affected configuration
- **Hardware:** NVIDIA Blackwell — verified on B300 (TP4, 1 node) and GB300NVL (TP4, 1 node).
- **Quantization:** `--quantization modelopt_fp4` (NVFP4), with the swizzled
  scale-factor layout path (`is_sf_swizzled_layout=True`).
- **Models verified:**
  - Llama-3.3-70B-Instruct, NVFP4 checkpoint (dense, TP4)
  - Kimi-K2.5, NVFP4 checkpoint (1T-A32B MoE, TP4, `--kv-cache-dtype fp8_e4m3`)
- **Software stack used in the failing automation runs** (as reported by the
  nightly automation): CUDA 13.0.2, PyTorch 2.11.0+cu130, FlashInfer 0.6.11.post2,
  Triton 3.6.0.

## The offending change

`create_fp4_scale_tensor` (`vllm/_custom_ops.py`) allocates the output scale
tensor for `scaled_fp4_quant`. In the swizzled branch it pads the scale tensor to
`(round_up(m, 128), round_up(n // 16, 4) // 4)` and stores it as `int32`. PR #42988
switched this allocation (and the non-swizzled `uint8` allocation) from
`torch.zeros` to `torch.empty`:

```python
    block_size = 16
    if is_sf_swizzled_layout:
        rounded_m = round_up(m, 128)
        scale_n = n // block_size
        rounded_n = round_up(scale_n, 4)
        return torch.empty(                       # was torch.zeros
            (rounded_m, rounded_n // 4), device=device, dtype=torch.int32
        )
    else:
        return torch.empty((m, n // block_size), device=device, dtype=torch.uint8)
```

## Root cause

The swizzled scale buffer is **padded in both dimensions**: rows are rounded up to a multiple of 128, and the packed-K dimension is rounded up to a multiple of 4.

The NVFP4 quantization kernel (`cvt_fp16_to_fp4` in `csrc/.../fp4/nvfp4_quant_kernels.cu`) writes scale factors to addresses computed by a swizzled-layout mapping (`cvt_quant_to_fp4_get_sf_out_offset` in `nvfp4_utils.cuh`), whose layout is `[numMTiles, numKTiles, 32, 4, 4]` with `numKTiles = (outputCols + 63) / 64`. The kernel's grid-stride loop iterates over padded rows/cols, and the in-kernel comment asserts it visits "every single scale factor address to initialize it.

In practice the quantization kernel does not write every element of this padded buffer that the consuming NVFP4 GEMM later reads. When the buffer is allocated with `torch.empty`, those unwritten elements contain whatever was previously resident in device memory, so the consumed scale factors are garbage, corrupting dequantization. The two observed symptoms (both measured below) are:
- the output sequence length collapsing far below the requested 1000 tokens (measured OSL 65–127), i.e. degenerate generation, and
- the per-token decode latency increasing several-fold (measured average inter-token latency ~50–98 ms, vs ~6–7 ms once fixed).

We did not enumerate exactly which elements are skipped. The claim we rely on is narrower and is directly proven by the fix: zero-initializing the buffer fully restores correct behavior, which is only possible if the kernel leaves consumer-visible elements unwritten. With `torch.zeros`, every unwritten element reads a defined `0` instead of garbage, the long-standing pre-PR behavior.

## Reproduction

The numbers below were produced with the same harness the automation uses: `vllm serve` + `aiperf profile`, single node, TP4, ISL=1000, OSL=1000, `ignore_eos:true`, 5 requests per concurrency + 1 warmup, parsed with the standard aiperf metrics parser.

Server (Llama example):

```bash
vllm serve <Llama-3.3-70B-Instruct-NVFP4> \
  --gpu-memory-utilization 0.9 \
  --served-model-name llama3.3-70b \
  --tensor-parallel-size 4 --pipeline-parallel-size 1 --data-parallel-size 1 \
  --max-model-len 2200 \
  --load-format runai_streamer \
  --quantization modelopt_fp4
```

Kimi adds `--kv-cache-dtype fp8_e4m3 --trust-remote-code` to the server, and `--tokenizer-trust-remote-code` to the `aiperf` client command below.

Client (per concurrency `C`):

```bash
aiperf profile --url localhost:$port --model <served_model_name> \
  --tokenizer <tokenizer> --endpoint-type completions --streaming --ui simple \
  --extra-inputs "min_tokens:1000" --extra-inputs "max_tokens:1000" \
  --extra-inputs "ignore_eos:true" \
  --concurrency $C \
  --synthetic-input-tokens-mean 1000 --synthetic-input-tokens-stddev 0 \
  --output-tokens-mean 1000 --output-tokens-stddev 0 \
  --request-count $((5 * C)) --warmup-request-count 1
```

To toggle the cause without recompiling, flip only the swizzled `int32` allocation between `torch.empty` (regressed) and `torch.zeros` (fixed) in the installed `vllm/_custom_ops.py` (`VLLM_DISABLE_COMPILE_CACHE=1` was set so a warm compile cache could not mask the toggle across in-session runs.)

### Result 1 — Per-hunk isolation on B300 (Llama-3.3-70B, con=1, ISL/OSL=1000/1000)

PR #42988 changed `torch.zeros`→`torch.empty` in several places. We reverted each independently to isolate which one is responsible. `OSL` is the average output sequence length (requested = 1000); it is derived as `avg_output_token_throughput / avg_request_throughput`. This derivation is approximate (it can land slightly above the 1000 `max_tokens` cap due to throughput/timing rounding); the meaningful signal is the contrast between the regressed runs (~65–127) and the healthy runs (~870–1040), not the exact value.

| Configuration                                            | Output tok/s | ITL (ms) | OSL    | Verdict        |
| -------------------------------------------------------- | -----------: | -------: | -----: | -------------- |
| Baseline (PR as-is, all `empty`)                         |    **19.56** |    50.62 |  127.0 | regressed      |
| Revert **all** `empty`→`zeros`                           |   **162.29** |     6.11 | 1041.4 | healthy        |
| Revert **fp8** util only                                 |        19.56 |    50.59 |  127.0 | no effect      |
| Revert **both** `_custom_ops` lines (swizzled + uint8)   |       146.43 |     6.73 |  943.8 | healthy        |
| Revert **swizzled `int32`** line only                    |   **135.40** |     7.35 |  873.2 | **healthy**    |
| Revert **non-swizzled `uint8`** line only                |        19.49 |    50.77 |  127.0 | no effect      |

The single swizzled `int32` allocation is necessary and sufficient: reverting only that line recovers throughput (19.56 → 135.40 tok/s), while reverting the non-swizzled `uint8` line or the fp8-util line has no effect.

### Result 2 — Before/after on GB300NVL (con=1, ISL/OSL=1000/1000)

| Model                | Allocation        | Output tok/s | ITL (ms) | OSL   |
| -------------------- | ----------------- | -----------: | -------: | ----: |
| Llama-3.3-70B (dense)| `empty` (regressed) |    **18.39** |    53.99 | 127.0 |
| Llama-3.3-70B (dense)| `zeros` (fixed)     |   **143.25** |     6.88 | 997.8 |
| Kimi-K2.5 (1T MoE)   | `empty` (regressed) |    **10.16** |    97.61 |  65.8 |
| Kimi-K2.5 (1T MoE)   | `zeros` (fixed)     |   **148.28** |     6.58 | 999.6 |

The regressed runs show the dual symptom: throughput collapse **and** output degeneration (OSL 127 and 65.8 vs. the requested 1000). The fix restores both throughput and full-length generation (OSL ≈ 1000).

### Result 3 — Fixed build scales cleanly across concurrency on GB300NVL

To confirm the fix does not break at load (and to directly address the concern about kernel/eval coverage raised during the original PR review), we swept concurrency 1/16/64/128 with the fixed (`zeros`) build:

| Concurrency | Llama tok/s | Llama ITL (ms) | Llama OSL | Kimi tok/s | Kimi ITL (ms) | Kimi OSL |
| ----------: | ----------: | -------------: | --------: | ---------: | ------------: | -------: |
| 1           |      137.93 |           7.15 |     995.0 |     147.57 |          6.62 |   1000.2 |
| 16          |     2020.14 |           7.64 |    1000.0 |    1047.85 |         13.68 |    999.9 |
| 64          |     7305.73 |           8.46 |     999.7 |    2770.39 |         22.10 |   1000.0 |
| 128         |    11478.84 |          10.73 |     999.8 |    4913.39 |         25.48 |   1000.0 |

Throughput scales monotonically, ITL stays bounded, and OSL stays at the full requested length (≈1000) at every concurrency for both models. No degeneration or collapse at scale.

> Note on run-to-run variance: the con=1 fixed Llama number is 143.25 tok/s in the
> dedicated single-concurrency run (Result 2) and 137.93 tok/s in the concurrency
> sweep (Result 3). These are independent 5-request runs; the small difference is
> ordinary measurement noise and is irrelevant next to the ~7–8× gap against the
> regressed baseline.

## Proposed fix

Restore `torch.zeros` for the swizzled `int32` scale allocation:

```diff
     block_size = 16
     if is_sf_swizzled_layout:
         rounded_m = round_up(m, 128)
         scale_n = n // block_size
         rounded_n = round_up(scale_n, 4)
-        return torch.empty(
+        return torch.zeros(
             (rounded_m, rounded_n // 4), device=device, dtype=torch.int32
         )
     else:
         return torch.empty((m, n // block_size), device=device, dtype=torch.uint8)
```

### Why

- `torch.zeros` is strictly a superset of `torch.empty`'s guarantees: it only adds initialization. Every element the kernel does write is still overwritten by the kernel; every element the kernel skips now reads a defined `0` instead of garbage. There is no code path where `zeros` is worse than `empty`.
- This restores the exact pre-PR behavior that ran correctly for a long time and that the kernel's contract implicitly relies on.
- The only cost is one memset on a small `int32` tensor. It is not measurable in throughput — in the B300 isolation run the fully-reverted (`zeros`) build measured 162.29 tok/s, i.e. no regression relative to the historical baseline.

A more complete long-term fix would be to make the quantization kernel write **all** scale-factor elements that the consumer reads (so the buffer need not be pre-zeroed). That is a larger, higher-risk kernel change; restoring `torch.zeros` is the minimal, safe fix for this regression.

## Regression window (from nightly automation history)

`con1_isl1000_osl1000_avg_output_token_throughput`:

- **Llama-3.3-70B:** last healthy at vLLM `0.21.1rc1.dev169+ga6682d1d2` (2026-05-22, 126.5 tok/s); first observed regressed at `0.21.1rc1.dev323+g1fc2cee50` (2026-05-28, 17.57 tok/s). Reported effect size −88.48%.
- **Kimi-K2.5:** last healthy at `0.21.1rc1.dev98+g6e889b582` (2026-05-26, 151.3 tok/s); first observed regressed at `0.21.1rc1.dev323+g1fc2cee50` (2026-05-28, 9.563 tok/s); persists through `0.22.1rc1.dev55+gf91fb2fcf`(2026-06-03, 10.14 tok/s). Reported effect size −93.36%.

## References

- **Offending PR:** [vllm-project/vllm#42988](https://github.com/vllm-project/vllm/pull/42988)
  — "[Perf] zeros -> empty to remove additional fill" (first bad commit `b29cbf0652`,
  first bad nightly wheel `0.21.1rc1.dev185+gb29cbf065`).
- **Affected source:** `vllm/_custom_ops.py::create_fp4_scale_tensor` (allocation), `csrc/.../quantization/fp4/nvfp4_quant_kernels.cu` (`cvt_fp16_to_fp4`), `csrc/.../quantization/fp4/nvfp4_utils.cuh` (`cvt_quant_to_fp4_get_sf_out_offset`).
- **Fix PR:** _to be linked once opened_ (one-line revert of the swizzled `int32` allocation to `torch.zeros`).

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.
